src/cls/rbd/cls_rbd_types.h

```diff
@@ -391,6 +411,82 @@ std::ostream& operator<<(std::ostream& os, const MirrorImageMap &image_map);
 
 WRITE_CLASS_ENCODER(MirrorImageMap);
 
+enum MigrationType {
+  MIGRATION_TYPE_UNDEFINED = 0,
+  MIGRATION_TYPE_SRC = 1,
+  MIGRATION_TYPE_DST = 2,
+};
+
+inline void encode(const MigrationType &type, bufferlist& bl) {
+  ::encode(static_cast<uint8_t>(type), bl);
+}
+
+inline void decode(MigrationType &type, bufferlist::iterator& it) {
+  uint8_t int_type;
+  ::decode(int_type, it);
+  type = static_cast<MigrationType>(int_type);
+}
+
+std::ostream& operator<<(std::ostream& os, const MigrationType& migration_type);
+
+enum MigrationState {
```

Comment:
Is this really needed? We know a migration is in-progress (via the v1 banner or v2 feature)

Reply:
I need started/completed state, so I could set completed on cleanup, before removing src image. This is needed for "purge snapshot" and "remove image" requests to succeed. Otherwise I would need some additional hack to make these procedures work with "migrating" set. Or remove "migrating" feature/banner before removal.